yarn-project/end-to-end/src/e2e_kernelless_simulation.test.ts

```diff
@@ -0,0 +1,199 @@
+import {
+  type AccountInterface,
+  AccountWallet,
+  AuthWitness,
+  type ContractArtifact,
+  Fr,
+  type Logger,
+  type PXE,
+  TxExecutionRequest,
+  type Wallet,
+  getContractInstanceFromDeployParams,
+} from '@aztec/aztec.js';
+import { AMMContract } from '@aztec/noir-contracts.js/AMM';
+import type { TokenContract } from '@aztec/noir-contracts.js/Token';
+import type { CompleteAddress, ContractInstanceWithAddress } from '@aztec/stdlib/contract';
+import type { SimulationOverrides, TxSimulationResult } from '@aztec/stdlib/tx';
+
+import { deployToken, mintTokensToPrivate } from './fixtures/token_utils.js';
+import { setup } from './fixtures/utils.js';
+
+export class CopyCatWallet extends AccountWallet {
+  constructor(
+    pxe: PXE,
+    account: AccountInterface,
+    private originalAddress: CompleteAddress,
+    private originalContractClassId: Fr,
+    private artifact: ContractArtifact,
+    private instance: ContractInstanceWithAddress,
+  ) {
+    super(pxe, account);
+  }
+
+  static async create(pxe: PXE, originalAccount: AccountWallet): Promise<CopyCatWallet> {
+    const simulatedAuthWitnessProvider = {
+      createAuthWit(messageHash: Fr): Promise<AuthWitness> {
+        return Promise.resolve(new AuthWitness(messageHash, []));
+      },
+    };
+    const nodeInfo = await pxe.getNodeInfo();
+    const originalAddress = originalAccount.getCompleteAddress();
+    const { contractInstance } = await pxe.getContractMetadata(originalAddress.address);
+    if (!contractInstance) {
+      throw new Error(`No contract instance found for address: ${originalAddress.address}`);
+    }
+    const { currentContractClassId: originalContractClassId } = contractInstance;
+    const accountInterface = new DefaultAccountInterface(simulatedAuthWitnessProvider, originalAddress, nodeInfo);
+    const { SimulatedAccountContractArtifact } = await import('@aztec/noir-contracts.js/SimulatedAccount');
+    const instance = await getContractInstanceFromDeployParams(SimulatedAccountContractArtifact, {});
+    return new CopyCatWallet(
+      pxe,
+      accountInterface,
+      originalAddress,
+      originalContractClassId,
+      SimulatedAccountContractArtifact,
+      instance,
+    );
+  }
+
+  override getCompleteAddress(): CompleteAddress {
+    return this.originalAddress;
+  }
+
+  override simulateTx(
+    txRequest: TxExecutionRequest,
+    simulatePublic: boolean,
+    skipTxValidation?: boolean,
+    skipFeeEnforcement?: boolean,
+    _overrides?: SimulationOverrides,
+  ): Promise<TxSimulationResult> {
+    const instanceOverrides = { [this.originalAddress.address.toString()]: this.instance };
+    const artifactOverrides = { [this.originalContractClassId.toString()]: this.artifact };
+    return this.pxe.simulateTx(txRequest, simulatePublic, skipTxValidation, skipFeeEnforcement, {
+      contracts: { instances: instanceOverrides, artifacts: artifactOverrides },
+    });
+  }
+}
+
+describe('Kernelless simulation', () => {
```

Comment:
Please document here what does this test test.

Reply:
Done